### ISSUE TITLE
Newbie questions and comments

### ISSUE THREAD

**Issue by edreamleo** [python/mypy#1769]
Hello all.  I have been following mypy with great interest.  Two comments, and then some questions.

C1. It's difficult to discover how to ask questions here.  Indeed, it is only in `CONTRIBUTING.md` that these words appear:

"Our main forum for discussion is the project's GitHub issue tracker. This is the right place to start a discussion of any of the above or most any other topic concerning the project."

Imo, these words belong on the main mypy page: http://mypy-lang.org/, not buried in instructions for devs.

C2. It seems strange to have to submit a new bug report in order to ask questions. Assuming mypy becomes popular, I would think it would be more convenient for you to create a Google discussion group.

Now the questions:

Q1. I'm confused about what happens when both a x.pyi and x.py exist.  Does mypy use both?

Q2. If I define an ivar like this in x.py:

```
aClass:
    aList = []
```

mypy complains:

    `error: Need type annotation for variable`

But if aList is assigned within a method, then mypy issues no warning.

Q3. I get the same warning if x.pyi contains, say:

```
aClass:
    aList = List[Any]
```

That is, the stub file's declaration for aList has no effect.  Otoh, if x.py contains:

```
aClass:
    aList = [] # type: List[str]
```

then mypy issues no warning.

In short, the .pyi file seems to have no effect.

Q4. If x.py contains:

```
aClass:
    aList = [] # type: List[Any]
```

Then mypy complains: `error: Name 'Any' is not defined` unless something like the following appears in x.py: `from typing import Any, Dict, List, Optional, Sequence, Tuple, Union`

This seems dubious.  It's one thing to add type comments to a file, it's another to require definitions of names used in those comments.  Or am I missing something?

Anyway, I'm glad to have found the place to ask questions, and I'm looking forward to using mypy.

Edward

**Comment by gvanrossum** [comment]
> C1. It's difficult to discover how to ask questions here. Indeed, it is only in CONTRIBUTING.md that these words appear:
> 
> "Our main forum for discussion is the project's GitHub issue tracker. This is the right place to start a discussion of any of the above or most any other topic concerning the project."
> 
> Imo, these words belong on the main mypy page: http://mypy-lang.org/, not buried in instructions for devs.

This policy is intentional, but you're right we should emphasize it more. I think it should also be added to README.md (in addition to mypy-lang.org).

> C2. It seems strange to have to submit a new bug report in order to ask questions. Assuming mypy becomes popular, I would think it would be more convenient for you to create a Google discussion group.

It is fully intentional. The GitHub tracker is quite usable as a mailing list substitute IMO. Using the tracker makes various discussion management/moderation practices easy that are hard to do using a traditional mailing list. Having a tracker places the focus on getting answers -- it's similar in idea to using StackOverflow.com.

I'll get to the rest later.

**Comment by gvanrossum** [comment]
> Q1. I'm confused about what happens when both a x.pyi and x.py exist. Does mypy use both?

It uses one or the other but not both. If they exist in the same
directory, mypy prefers x.pyi.

If x.py is found on the mypy search path in an earlier directory than
x.pyi, x.py wins.

The mypy search patch is determined by:
- $MYPYPATH
- Source directories determined by the files passed on the command line [1]
- The relevant directories in typeshed

[1] This is further complicated by crawling up until no __init__.py[i]
is found, but that just makes it more useful (just harder to specify
in one sentence).

> Q2. If I define an ivar like this:
>
> aClass:
>     aList = []
>
> mypy complains:
>
> `error: Need type annotation for variable`

That's because it doesn't know the type of the list items.

> But if aList is assigned within a method, then mypy issues no warning.

I presume you're referring to a local variable initialized to an empty
list (or other container). In that case mypy applies some heuristics
based on nearby method calls on that object, e.g. it understands this:
```
a = []
for i in range(10):
    a.append(i*3)
```
However it is still possible to confound it, and then you'll still get
the "Need type annotation for variable" error message.

> Q3. I get the same warning if x.pyi contains, say:
>
> aClass:
>     aList = List[Any]
>
> That is, the stub file's declaration for aList has no effect.

This seems a bug. List[Any] is a type -- so `aList = List[Any]` makes
aList a type alias. But type aliases should only occur at the top
level.

Unless you meant to write
```
    aList = ...  # type: List[Any]
```

> Otoh, if x.py contains:
>
> aClass:
>     aList = [] # type: List[str]
>
> then mypy issues no warning.

Correct.

> In short, the .pyi file seems to have no effect.

Maybe the answer to Q1 is relevant -- x.py and x.pyi are never both processed.

> Q4. If x.py contains:
>
> aClass:
>     aList = [] # type: List[Any]
>
> Then mypy complains: error: Name 'Any' is not defined unless something like the following appears in x.py: from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
>
> This seems dubious. It's one thing to add type comments to a file, it's another to require definitions of names used in those comments. Or am I missing something?

How would mypy know where those names are defined? It doesn't augment
the builtins, and IMO it shouldn't. It processes type annotations
(whether in comments or in PEP-3107-style function annotations) using
the same symbol table as it uses to process the rest of your module.

**Comment by edreamleo** [comment]
On Wed, Jun 29, 2016 at 7:43 PM, Guido van Rossum <notifications@github.com>
wrote:

> > Q1. I'm confused about what happens when both a x.pyi and x.py exist.
> Does mypy use both?
>
> It uses one or the other but not both. If they exist in the same
> directory, mypy prefers x.pyi.
>

​Thanks.  Probably should be mentioned here
<https://mypy.readthedocs.io/en/latest/command_line.html>.

>
> [1] This is further complicated by crawling up until no __init__.py[i]
> is found, but that just makes it more useful (just harder to specify
> in one sentence).
>
> > Q2. If I define an ivar like this:
> >
> ​​
> > aClass:
> > aList = []
> >
> > mypy complains:
> >
> > `error: Need type annotation for variable`
>
> That's because it doesn't know the type of the list items.
>
> > But if aList is assigned within a method, then mypy issues no warning.
>
> I presume you're referring to a local variable initialized to an empty
> list (or other container).


​No.  I was referring to an ivar defined outside of any method. For
example, given this test file:

```
class aClass:
    aList = []
    def foo(self):
        print(self.aList)

class aClass2:
    def __init__(self):
        self.aList = []
    def foo(self):
        print(self.aList)

i1, i2 = aClass(), aClass2()
i1.foo()
i2.foo()
```

the output is:

```
test.py: note: In class "aClass":
test.py:2: error: Need type annotation for variable
```

That is, mypy complains about the definition in aClass, but not aClass2.

> Q4. If x.py contains:
> >
> > aClass:
> > aList = [] # type: List[Any]
> >
> > Then mypy complains: error: Name 'Any' is not defined unless something
> like the following appears in x.py: from typing import Any, Dict, List,
> Optional, Sequence, Tuple, Union
> >
> > This seems dubious. It's one thing to add type comments to a file, it's
> another to require definitions of names used in those comments. Or am I
> missing something?
>
> How would mypy know where those names are defined?


​I don't much care, just so long as it knows about the standard types that
are going to be ubiquitous in the type comments.  But I see your point:
even if a special case if made for the types in the typing module (what the
mypy documentation presently calls the builtin types
<https://mypy.readthedocs.io/en/latest/builtin_types.html>), the user will
probably use other types that mypy won't know about.

Hmm.  Thinking out loud here.  I don't mind type hint comments all that
much, but I am uncomfortable having to put new imports into checked files,
especially 'from typing import *'.  Would it be reasonable to augment
imports in x.py with those from x.pyi?  As I write this, I would guess not,
but you can see what I'm trying to do...

Edward

**Comment by gvanrossum** [comment]
I added a very prominent note about using the tracker for questions to the top of README.md, and added a section describing the import search algorithm to the command line docs.

> Q3 [...] That is, mypy complains about the definition in aClass, but not aClass2.

Oh, that's a beginner's mistake. Neither `__init__` not `foo` has a type annotation, so mypy simply doesn't look inside them. If you add `-> None` to each method you'll see that it complains about both. The difference is that mypy _always_ analyzes toplevel code and code in toplevel classes; but it only analyzes code inside functions/methods that are annotated. That's because of gradual typing -- in a typical large program that's mostly unannotated, you would get a large number of errors about unannotated methods, and that would be unhelpful. (If you disagree, just pass `--check-untyped-defs` and you can see for youself. :-)

**Comment by gvanrossum** [comment]
> Q4 [...]

Closing the loop here, mypy's design is intended to make the most sense for Python 3 -- Python 2 support is designed to be as similar as possible while not requiring changes to the Python 2 syntax. In Python 3, signature annotations are required (by PEP 3107) to be evaluated where they occur, so the names they reference must be imported. Since this means that in general Python 3 requires the imports, we've decided to require them for Python 2 as well. In part that's because that's the simplest implementation given that mypy started out only supporting Python 3; in part it's so that you'll be prepared when you start porting your code to Python 3 (which is actually an important use case for mypy).

**Comment by edreamleo** [comment]
On Thu, Jun 30, 2016 at 4:09 PM, Guido van Rossum <notifications@github.com>
wrote:

> I added a very prominent note about using the tracker for questions to the
> top of README.md, and added a section describing the import search
> algorithm to the command line docs.
>
​Thanks.​


> Q3 [...] That is, mypy complains about the definition in aClass, but not
> aClass2.
>
> Oh, that's a beginner's mistake. Neither __init__ not foo has a type
> annotation, so mypy simply doesn't look inside them.
>
​I see.  Thanks for the correction.​


​Edward

**Comment by edreamleo** [comment]
On Thu, Jun 30, 2016 at 4:14 PM, Guido van Rossum <notifications@github.com>
wrote:

> Q4 [...]
>
> Closing the loop here, mypy's design is intended to make the most sense
> for Python 3 -- Python 2 support is designed to be as similar as possible
> while not requiring changes to the Python 2 syntax. In Python 3, signature
> annotations are required (by PEP 3107) to be evaluated where they occur, so
> the names they reference must be imported. Since this means that in general
> Python 3 requires the imports, we've decided to require them for Python 2
> as well. In part that's because that's the simplest implementation given
> that mypy started out only supporting Python 3; in part it's so that you'll
> be prepared when you start porting your code to Python 3 (which is actually
> an important use case for mypy).
>
​Seems reasonable.​

​ Many thanks for your patient explanations. I think we can close this
discussion.

Edward

**Comment by gvanrossum** [comment]
Glad it all makes sense! And thanks for the push to improve our docs.
